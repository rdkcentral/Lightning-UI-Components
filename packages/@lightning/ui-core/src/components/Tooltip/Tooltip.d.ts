import lng from '@lightningjs/core';
import Base from '../Base';
import type { Color, StylePartial } from '../../types/lui';
import type { TextBoxStyles } from '../TextBox';
import Tooltip, { TooltipStyles } from './Tooltip';

type TransitionObject = {
  delay: number;
  duration: number;
  timingFunction: string;
};

export type TooltipStyles = {
  backgroundColor: Color;
  marginBottom: number;
  paddingX: number;
  paddingY: number;
  pointerW: number;
  pointerH: number;
  radius: lng.Tools.CornerRadius;
  textStyle: TextBoxStyles;
  transition: TransitionObject;
};

export default class Tooltip extends Base {
  title?: string;
  delayVisible?: number;
  timeVisible?: number;
  get style(): TooltipStyles;
  set style(v: StylePartial<TooltipStyles>);
}