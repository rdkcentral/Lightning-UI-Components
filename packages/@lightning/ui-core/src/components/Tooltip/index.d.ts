import Tooltip, { TooltipStyles } from './Tooltip';

export { Tooltip as default, TooltipStyles };
