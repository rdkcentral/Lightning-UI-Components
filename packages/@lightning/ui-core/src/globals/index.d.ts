export { default as context } from './context';
