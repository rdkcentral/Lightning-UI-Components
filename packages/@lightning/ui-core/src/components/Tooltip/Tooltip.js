import { Bubble } from '../../textures/index.js';
import TextBox from '../TextBox/index.js';
import Base from '../Base/index.js';
import * as styles from './Tooltip.styles.js';

export default class Tooltip extends Base {
  static get __componentName() {
    return 'Tooltip';
  }

  static get __themeStyle() {
    return styles;
  }

  static _template() {
    return {
      alpha: 0,
      scale: 0.5,
      mountX: 0.5,
      x: w => w / 2,
      Background: {
        Text: {
          type: TextBox,
          signals: {
            textBoxChanged: '_textLoaded'
          }
        }
      }
    };
  }

  static get properties() {
    return ['title', 'delayVisible', 'timeVisible'];
  }

  static get tags() {
    return ['Background', { name: 'Text', path: 'Background.Text' }];
  }

  _update() {
    this._updateText();
  }

  _updateText() {
    if (this._Text) {
      this._Text.patch({
        content: this.title,
        style: { textStyle: this.style.textStyle }
      });
    }
  }

  _textLoaded() {
    this._updateBackground();
    this._updateTextPosition();
  }

  _updateBackground() {
    const backgroundH =
      this._Text.finalH + this.style.paddingY * 2 + this.style.pointerH;
    const backgroundW = this._Text.finalW + this.style.paddingX * 2;

    this.patch({
      w: backgroundW,
      h: backgroundH,
      mountY: 1,
      y: -this.style.marginBottom,
      Background: {
        w: backgroundW,
        h: backgroundH,
        texture: {
          type: Bubble,
          w: backgroundW,
          h: backgroundH,
          radius: this.style.radius,
          pointerW: this.style.pointerW,
          pointerH: this.style.pointerH,
          color: this.style.backgroundColor
        }
      }
    });
  }

  _updateTextPosition() {
    if (this._Text) {
      this._Text.patch({
        mount: 0.5,
        x: this._Background.w / 2,
        y: (this._Background.h - this.style.pointerH) / 2
      });
    }
  }

  _clearTimers() {
    clearTimeout(this._hideTimer);
    clearTimeout(this._showTimer);
  }

  _transitionIn() {
    const smooth = {
      smooth: {
        alpha: [1, this.style.transition],
        scale: [1, this.style.transition]
      }
    };

    if (this.delayVisible) {
      this._showTimer = setTimeout(() => {
        this.patch({ smooth });
      }, this.delayVisible);
    } else {
      this._showTimer = undefined;
      this.patch({ smooth });
    }

    this._hideTimer = this.timeVisible
      ? setTimeout(() => {
          this._unfocus();
        }, this.timeVisible + (this.delayVisible || 0))
      : undefined;
  }

  _transitionOut() {
    this.patch({
      smooth: {
        alpha: [0, this.style.transition],
        scale: [0.5, this.style.transition]
      }
    });
  }

  _focus() {
    this._clearTimers();
    this._transitionIn();
  }

  _unfocus() {
    this._clearTimers();
    this._transitionOut();
  }

  set announce(announce) {
    super.announce = announce;
  }

  get announce() {
    return this._announce || (this._Text && this._Text.announce);
  }
}
