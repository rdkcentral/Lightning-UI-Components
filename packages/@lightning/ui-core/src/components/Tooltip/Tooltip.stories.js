import lng from '@lightningjs/core';
import Button from '../Button/index.js';
import { createModeControl } from '../../../storybook/index.js';
import mdx from './Tooltip.mdx';
import Tooltip from './index.js';
import { CATEGORIES } from 'lightning-ui-docs';

export default {
  title: `${CATEGORIES[16]}/Tooltip`,
  args: {
    title: 'Tooltip',
    delayVisible: 0,
    timeVisible: 0
  },
  argTypes: {
    ...createModeControl(['focused'], 'focused'),
    title: { control: 'text' },
    delayVisible: { control: 'number' },
    timeVisible: { control: 'number' }
  },
  parameters: {
    docs: {
      page: mdx
    }
  }
};

export const Basic = () =>
  class Basic extends lng.Component {
    static _template() {
      return {
        Tooltip: {
          type: Tooltip
        }
      };
    }
  };

export const LongTitle = Basic.bind({});
LongTitle.args = {
  x: 1280 / 2,
  title:
    'This is a long message. Text will remain on a single line and does not have a maximum width'
};

export const WithButton = args =>
  class WithButton extends lng.Component {
    static _template() {
      return {
        Button: {
          type: class extends Button {
            _getFocused() {
              return this.childList.getByRef('Tooltip');
            }
          },
          x: 200,
          y: 100,
          title: 'Button',
          backgroundType: 'stroke',
          Tooltip: {
            type: Tooltip,
            title: args.title,
            delayVisible: args.delayVisible,
            timeVisible: args.timeVisible
          }
        }
      };
    }
  };

WithButton.args = {
  bottomMargin: 24
};

WithButton.parameters = {
  argActions: {
    title: (title, component) => {
      component.tag('Button.Tooltip').title = title;
    },
    delayVisible: (delayVisible, component) => {
      component.tag('Button.Tooltip').delayVisible = delayVisible;
    },
    timeVisible: (timeVisible, component) => {
      component.tag('Button.Tooltip').timeVisible = timeVisible;
    }
  }
};
