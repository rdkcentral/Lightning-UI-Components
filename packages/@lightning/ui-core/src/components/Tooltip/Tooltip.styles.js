export const base = theme => ({
  marginBottom: theme.spacer.xxxl,
  paddingX: theme.spacer.lg,
  paddingY: theme.spacer.md,
  pointerW: theme.spacer.xl,
  pointerH: theme.spacer.md + theme.spacer.xs,
  radius: theme.radius.md,
  textStyle: {
    ...theme.typography.caption1,
    textColor: theme.color.textInverse
  },
  transition: theme.animation.utility // TODO: should delay and duration be xfast?
});

export const tone = theme => ({
  neutral: {
    backgroundColor: theme.color.fillNeutral,
    textStyle: { textColor: theme.color.textInverse }
  },
  inverse: {
    backgroundColor: theme.color.fillInverse,
    textStyle: { textColor: theme.color.textNeutral }
  },
  brand: {
    backgroundColor: theme.color.fillBrand,
    textStyle: { textColor: theme.color.textNeutral }
  }
});
