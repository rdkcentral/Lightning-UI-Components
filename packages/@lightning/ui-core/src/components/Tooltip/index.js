export { default as default } from './Tooltip.js';
