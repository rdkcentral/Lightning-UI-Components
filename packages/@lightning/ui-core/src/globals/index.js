export { default as context } from './context/index.js';
export { updateManager } from './global-update-manager/GlobalUpdateManager.js';
export { default as pool } from './pool/index.js';
