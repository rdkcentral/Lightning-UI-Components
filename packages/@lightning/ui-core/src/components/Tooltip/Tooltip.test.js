import {
  makeCreateComponent,
  completeAnimation
} from '@lightning/ui-test-utils';
import Tooltip from './index.js';
import { jest } from '@jest/globals';

const createTooltip = makeCreateComponent(Tooltip, {}, { focused: false });

describe('Tooltip', () => {
  let tooltip, testRenderer;

  beforeEach(() => {
    [tooltip, testRenderer] = createTooltip({ title: 'Tooltip' });
  });

  it('renders', async () => {
    testRenderer.focus();
    await completeAnimation(tooltip, ['alpha', 'scale']);
    const tree = testRenderer.toJSON(2);
    expect(tree).toMatchSnapshot();
  });

  it("sets the announce string to the tooltip's title", () => {
    const title = 'Helper text';
    tooltip.title = title;
    // need to focus to make the component render in, otherwise there is no TextBox rendered
    tooltip._focus();
    testRenderer.forceAllUpdates();
    expect(tooltip.announce).toBe(title);
  });

  it('overrides the announce string', () => {
    const overrideString = 'Custom announce string';
    tooltip.announce = overrideString;
    testRenderer.forceAllUpdates();
    expect(tooltip.announce).toBe(overrideString);
  });

  describe('visibility', () => {
    it('should be visible when focused', async () => {
      expect(tooltip.alpha).toEqual(0);
      expect(tooltip.scale).toBe(0.5);

      tooltip._focus();
      await completeAnimation(tooltip, ['alpha', 'scale']);

      expect(tooltip.alpha).toEqual(1);
      expect(tooltip.scale).toEqual(1);
    });

    it('should be hidden when unfocused', async () => {
      [tooltip, testRenderer] = createTooltip(
        { title: 'Tooltip' },
        { focused: true }
      );
      await completeAnimation(tooltip, ['alpha', 'scale']);

      expect(tooltip.alpha).toEqual(1);
      expect(tooltip.scale).toEqual(1);

      tooltip._unfocus();
      await completeAnimation(tooltip, ['alpha', 'scale']);

      expect(tooltip.alpha).toEqual(0);
      expect(tooltip.scale).toBe(0.5);
    });

    it('should show after delay if delayVisible is set', async () => {
      jest.useFakeTimers();
      jest.spyOn(global, 'setTimeout');

      const delayVisible = 5000;
      [tooltip, testRenderer] = createTooltip({ delayVisible });

      expect(tooltip.alpha).toEqual(0);
      expect(tooltip.scale).toBe(0.5);
      expect(tooltip._showTimer).toBeUndefined();
      expect(setTimeout).not.toHaveBeenCalledWith(
        expect.any(Function),
        delayVisible
      );

      tooltip._focus();

      // should remain hidden
      expect(tooltip.alpha).toEqual(0);
      expect(tooltip.scale).toBe(0.5);
      expect(tooltip._showTimer).not.toBeUndefined();
      expect(setTimeout).toHaveBeenCalledWith(
        expect.any(Function),
        delayVisible
      );

      jest.advanceTimersByTime(delayVisible);
      testRenderer.update();

      // should show after timer runs
      expect(tooltip.transition('alpha').targetValue).toEqual(1);
      expect(tooltip.transition('scale').targetValue).toEqual(1);
    });

    it('should hide after timer ends if timeVisible is set', async () => {
      jest.useFakeTimers();
      jest.spyOn(global, 'setTimeout');

      const timeVisible = 5000;
      expect(tooltip._hideTimer).toBeUndefined();

      tooltip._focus();
      testRenderer.update();

      expect(setTimeout).not.toHaveBeenCalled();
      expect(tooltip._hideTimer).toBeUndefined();
      expect(tooltip.transition('alpha').targetValue).toEqual(1);
      expect(tooltip.transition('scale').targetValue).toEqual(1);

      tooltip.timeVisible = timeVisible;

      expect(tooltip._hideTimer).toBeUndefined();

      tooltip._focus();
      testRenderer.update();

      expect(setTimeout).toHaveBeenCalledWith(
        expect.any(Function),
        timeVisible
      );
      expect(tooltip._hideTimer).not.toBeUndefined();
      expect(tooltip.transition('alpha').targetValue).toEqual(1);
      expect(tooltip.transition('scale').targetValue).toEqual(1);

      // timer started on focus and hide after timer is called
      jest.advanceTimersByTime(timeVisible);
      testRenderer.update();

      expect(tooltip.transition('alpha').targetValue).toEqual(0);
      expect(tooltip.transition('scale').targetValue).toEqual(0.5);
    });
  });

  describe('dimensions', () => {
    beforeEach(async () => {
      jest.useFakeTimers();
      jest.spyOn(global, 'setTimeout');

      tooltip._focus();
      testRenderer.update();
    });

    it('should give the tooltip a margin bottom', () => {
      expect(tooltip.y).toBeLessThan(0);
    });

    it('should grow in width for longer text', () => {
      const initialW = tooltip._Background.w;
      tooltip.title = 'Longer Tooltip';

      testRenderer.forceAllUpdates();
      jest.runAllTimers();
      testRenderer.update();

      expect(tooltip._Background.w).toBeGreaterThan(initialW);
    });
  });
});
