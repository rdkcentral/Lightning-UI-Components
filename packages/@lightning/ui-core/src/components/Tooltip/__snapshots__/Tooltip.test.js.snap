// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tooltip visibility should show after delay if delayVisible is set 1`] = `
{
  "Component": {
    "Background": {
      "active": false,
      "alpha": 1,
      "attached": true,
      "boundsMargin": null,
      "children": {
        "Text": {
          "active": false,
          "alpha": 0.001,
          "attached": true,
          "boundsMargin": null,
          "clipping": false,
          "color": 4294967295,
          "enabled": false,
          "flex": false,
          "flexItem": false,
          "h": 0,
          "hasFinalFocus": false,
          "hasFocus": false,
          "isComponent": true,
          "mount": 0,
          "mountX": 0,
          "mountY": 0,
          "pivot": 0.5,
          "pivotX": 0.5,
          "pivotY": 0.5,
          "ref": "Text",
          "renderOfScreen": undefined,
          "renderToTexture": false,
          "scale": 1,
          "scaleX": 1,
          "scaleY": 1,
          "state": "",
          "tag": [Function],
          "tags": [
            "Text",
          ],
          "type": "TextBox",
          "visible": true,
          "w": 0,
          "x": 0,
          "y": 0,
          "zIndex": 0,
        },
      },
      "clipping": false,
      "color": 4294967295,
      "enabled": false,
      "flex": false,
      "flexItem": false,
      "h": 0,
      "isComponent": undefined,
      "mount": 0,
      "mountX": 0,
      "mountY": 0,
      "pivot": 0.5,
      "pivotX": 0.5,
      "pivotY": 0.5,
      "ref": "Background",
      "renderOfScreen": undefined,
      "renderToTexture": false,
      "scale": 1,
      "scaleX": 1,
      "scaleY": 1,
      "state": undefined,
      "tag": [Function],
      "tags": [
        "Background",
      ],
      "visible": true,
      "w": 0,
      "x": 0,
      "y": 0,
      "zIndex": 0,
    },
  },
}
`;
